proposals/NNNN-library-evolution.md

```diff
@@ -0,0 +1,374 @@
+
+The goal of this proposal is to reduce the burden on library developers by building into the compiler an automatic mechanism to reserve the flexibility to alter the internal representation of structs and enums, without manual workarounds. This mechanism can be implemented in such a way that optimizations such as stack allocation or contiguous inline storage of structs can still happen, while leaving the size of the type to be determined at runtime. 
+
+  [SE-0192]: https://github.com/apple/swift-evolution/blob/master/proposals/0192-non-exhaustive-enums.md
+
+## Proposed solution
+
+The compiler will gain a new mode, called "library evolution mode". This mode will be off by default. A library compiled with this mode enabled is said to be _ABI-stable_.
+
+When a library is compiled with library evolution mode enabled, it is not an ABI-breaking change to modify the fields in a struct (to add, remove, or reorder them), and to add new enum cases (including with associated values). This implies that clients must manipulate fields and enum cases indirectly, via non-inlable function calls. Information such as the size of the type, and the layout of its fields, becomes something that can only be determined at runtime. Types that reserve this flexibility are referred to as "resilient types".
+
+A new `@frozen` keyword will be introduced to allow library authors to opt out of this flexibility on a per-type basis. This keyword promises that stored instance properties within the struct will not be *added,* *removed,* or *reordered*. The compiler will use this for optimization purposes when compiling clients of the struct. The precise set of allowed changes is defined below.
+
+This attribute has no effect when binary stability mode is off. In that case, the compiler acts as if *every* struct or enum in the library being compiled is "frozen".
+
+## Detailed design
+
+### Library evolution mode
+
+A new command-line argument, `-enable-library-evolution`, will enable this new mode.
+
+Turning on library evolution mode will have the following affects on source code:
+
+- The default behavior for enums will change to be non-frozen. This is the only change visible to _users_ of the library, who will need to use the `@unknown default:` technique described in SE-0192.
+- To ensure that all fields are initialized, inlinable `init`s of the type must delegate to a non-inlinable `init`.
+
+### `@frozen` on `struct` types
+
+When a library author is certain that there will never be a need to add fields to a struct in future, they may mark that type as `@frozen`. 
+
+This will allow the compiler to optimize away at compile time some calls that would otherwise need to be made at runtime (for example, it might access ).
+
+When compiling with binary stability mode on, a struct can be marked `@frozen` as long as it meets all of the following conditions:
+
+- The struct is *ABI-public* (see [SE-0193][]), i.e. `public` or marked `@abiPublic`.
```

Comment:
This became `@usableFromInline`, and I forget if we left the "ABI-public" term in SE-0193 when it did.

Reply:
Yeah that term's used there still.